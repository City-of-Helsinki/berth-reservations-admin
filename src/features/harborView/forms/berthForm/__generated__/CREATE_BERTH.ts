/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { CreateBerthMutationInput } from "./../../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL mutation operation: CREATE_BERTH
// ====================================================

export interface CREATE_BERTH_createBerth {
  __typename: "CreateBerthMutationPayload";
  clientMutationId: string | null;
}

export interface CREATE_BERTH {
  createBerth: CREATE_BERTH_createBerth | null;
}

export interface CREATE_BERTHVariables {
  input: CreateBerthMutationInput;
}
