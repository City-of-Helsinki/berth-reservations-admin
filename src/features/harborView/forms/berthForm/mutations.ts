import gql from 'graphql-tag';

export const UPDATE_BERTH_MUTATION = gql`
  mutation UPDATE_BERTH($input: UpdateBerthMutationInput!) {
    updateBerth(input: $input) {
      clientMutationId
    }
  }
`;

export const DELETE_BERTH_MUTATION = gql`
  mutation DELETE_BERTH($input: DeleteBerthMutationInput!) {
    deleteBerth(input: $input) {
      clientMutationId
    }
  }
`;

export const CREATE_BERTH_MUTATION = gql`
  mutation CREATE_BERTH($input: CreateBerthMutationInput!) {
    createBerth(input: $input) {
      clientMutationId
    }
  }
`;
