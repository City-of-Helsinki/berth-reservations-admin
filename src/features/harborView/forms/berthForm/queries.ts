import gql from 'graphql-tag';

export const INDIVIDUAL_BERTH_QUERY = gql`
  query INDIVIDUAL_BERTH($id: ID!) {
    berth(id: $id) {
      number
      comment
      isActive
      pier {
        id
        properties {
          identifier
        }
      }
      mooringType
      width
      length
      depth
    }
  }
`;
