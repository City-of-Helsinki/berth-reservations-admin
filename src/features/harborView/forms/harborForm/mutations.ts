import { gql } from 'apollo-boost';

export const UPDATE_HARBOR_MUTATION = gql`
  mutation UPDATE_HARBOR($input: UpdateHarborMutationInput!) {
    updateHarbor(input: $input) {
      clientMutationId
    }
  }
`;
