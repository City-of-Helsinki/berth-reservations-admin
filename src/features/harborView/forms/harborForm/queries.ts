import { gql } from 'apollo-boost';

export const HARBOR_FORM_QUERY = gql`
  query HARBOR_FORM($id: ID!) {
    harbor(id: $id) {
      id
      properties {
        name
        streetAddress
        zipCode
        municipality
        wwwUrl
        imageFile
        maps {
          id
          url
        }
      }
    }
  }
`;
