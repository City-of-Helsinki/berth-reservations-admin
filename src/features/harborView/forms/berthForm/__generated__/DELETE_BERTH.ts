/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { DeleteBerthMutationInput } from "./../../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL mutation operation: DELETE_BERTH
// ====================================================

export interface DELETE_BERTH_deleteBerth {
  __typename: "DeleteBerthMutationPayload";
  clientMutationId: string | null;
}

export interface DELETE_BERTH {
  deleteBerth: DELETE_BERTH_deleteBerth | null;
}

export interface DELETE_BERTHVariables {
  input: DeleteBerthMutationInput;
}
