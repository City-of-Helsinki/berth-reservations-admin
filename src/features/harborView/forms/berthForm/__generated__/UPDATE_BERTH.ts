/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { UpdateBerthMutationInput } from "./../../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL mutation operation: UPDATE_BERTH
// ====================================================

export interface UPDATE_BERTH_updateBerth {
  __typename: "UpdateBerthMutationPayload";
  clientMutationId: string | null;
}

export interface UPDATE_BERTH {
  updateBerth: UPDATE_BERTH_updateBerth | null;
}

export interface UPDATE_BERTHVariables {
  input: UpdateBerthMutationInput;
}
