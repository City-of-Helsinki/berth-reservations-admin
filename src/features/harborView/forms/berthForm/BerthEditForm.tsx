import React from 'react';
import { useMutation, useQuery } from '@apollo/react-hooks';
import { useTranslation } from 'react-i18next';

import { INDIVIDUAL_BERTH_QUERY } from './queries';
import LoadingSpinner from '../../../../common/spinner/LoadingSpinner';
import { INDIVIDUAL_BERTH } from './__generated__/INDIVIDUAL_BERTH';
import { UPDATE_BERTH, UPDATE_BERTHVariables as UPDATE_BERTH_VARS } from './__generated__/UPDATE_BERTH';
import { DELETE_BERTH_MUTATION, UPDATE_BERTH_MUTATION } from './mutations';
import { Berth, FormProps } from '../types';
import { DELETE_BERTH, DELETE_BERTHVariables as DELETE_BERTH_VARS } from './__generated__/DELETE_BERTH';
import { getBerth } from './utils/utils';
import BerthForm from './BerthForm';
import { Pier } from '../../utils/utils';

interface BerthEditFormProps extends Omit<FormProps<Berth>, 'initialValues' | 'onCreate'> {
  berthId: string;
  pierOptions: Pier[];
}

const BerthEditForm: React.FC<BerthEditFormProps> = ({
  berthId,
  onCancel,
  onSubmit,
  onDelete,
  refetchQueries,
  pierOptions,
}) => {
  const { loading, error, data } = useQuery<INDIVIDUAL_BERTH>(INDIVIDUAL_BERTH_QUERY, { variables: { id: berthId } });

  const [updateBerth, { loading: isSubmitting }] = useMutation<UPDATE_BERTH, UPDATE_BERTH_VARS>(UPDATE_BERTH_MUTATION, {
    refetchQueries: [...(refetchQueries ?? []), { query: INDIVIDUAL_BERTH_QUERY, variables: { id: berthId } }],
  });

  const [deleteBerth] = useMutation<DELETE_BERTH, DELETE_BERTH_VARS>(DELETE_BERTH_MUTATION, {
    refetchQueries: refetchQueries ?? [],
  });

  const { t } = useTranslation();

  if (loading) return <LoadingSpinner isLoading={loading} />;
  if (error) return <div>{t('forms.common.error')}</div>;

  const initialValues = getBerth(data);

  return (
    <BerthForm
      initialValues={initialValues}
      isEditing={true}
      onCancel={onCancel}
      onDelete={(values) => deleteBerth({ variables: { input: { id: berthId } } }).then(() => onDelete?.(values))}
      onSubmitText={t('forms.common.update')}
      onSubmit={(values) => {
        const { number, pierId, width, length, mooringType, comment, isActive, depth } = values;
        updateBerth({
          variables: {
            input: {
              id: berthId,
              pierId,
              number,
              width,
              length,
              mooringType,
              comment,
              isActive,
              depth,
            },
          },
        }).then(() => onSubmit?.(values));
      }}
      isSubmitting={isSubmitting}
      pierOptions={pierOptions}
    />
  );
};

export default BerthEditForm;
