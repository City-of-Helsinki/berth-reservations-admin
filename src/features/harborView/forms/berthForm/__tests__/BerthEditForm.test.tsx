import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import waitForExpect from 'wait-for-expect';
import { act } from 'react-dom/test-utils';

import BerthEditForm from '../BerthEditForm';
import { INDIVIDUAL_BERTH_QUERY } from '../queries';
import { UPDATE_BERTH_MUTATION } from '../mutations';
import LoadingSpinner from '../../../../../common/spinner/LoadingSpinner';
import { BerthMooringType, PriceTier } from '../../../../../@types/__generated__/globalTypes';
import { Pier } from '../../../types';

const pierOptions: Pier[] = [
  {
    id: 'a',
    identifier: 'A',
    electricity: true,
    wasteCollection: true,
    water: true,
    lighting: true,
    gate: true,
    priceTier: PriceTier.TIER_1,
    suitableBoatTypes: [],
  },
];
const queryMock = {
  request: { query: INDIVIDUAL_BERTH_QUERY, variables: { id: 'a' } },
  result: {
    data: {
      berth: {
        __typename: 'BerthNode',
        id: 'MOCK-BERTH',
        number: '2',
        comment: '',
        isActive: true,
        pier: {
          __typename: 'PierNode',
          id: pierOptions[0].id,
          properties: {
            __typename: 'PierProperties',
            identifier: pierOptions[0].identifier,
          },
        },
        mooringType: BerthMooringType.SINGLE_SLIP_PLACE,
        width: 2.25,
        length: 5,
        depth: 1,
      },
    },
  },
};

describe('features/harborView/BerthEditForm', () => {
  const waitForContent = async (wrapper: ReactWrapper) => {
    await act(async () => {
      await waitForExpect(() => {
        wrapper.update();
        expect(wrapper.contains(<LoadingSpinner />)).toBeFalsy();
      });
    });
  };

  it('initially renders loading spinner', () => {
    const wrapper = mount(
      <MockedProvider mocks={[queryMock]}>
        <BerthEditForm berthId="a" pierOptions={pierOptions} />
      </MockedProvider>
    );
    expect(wrapper.contains(<LoadingSpinner />)).toBeTruthy();
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('renders content after loading', async () => {
    const wrapper = mount(
      <MockedProvider mocks={[queryMock]}>
        <BerthEditForm berthId="a" pierOptions={pierOptions} />
      </MockedProvider>
    );
    await waitForContent(wrapper);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('calls update mutation on save', async () => {
    let updateMockCalled = false;
    const onUpdateMock = jest.fn();
    const updateMock = {
      request: {
        query: UPDATE_BERTH_MUTATION,
        variables: {
          input: {
            id: 'a',
            number: '2',
            pierId: pierOptions[0].id,
            width: 2.25,
            length: 5,
            depth: 1,
            mooringType: 'SINGLE_SLIP_PLACE',
            comment: '',
            isActive: true,
          },
        },
      },
      result: () => {
        updateMockCalled = true;
        return {
          data: { updateBerth: { clientMutationId: '-', __typename: 'ID' } },
        };
      },
    };
    const wrapper = mount(
      // We need queryMock twice here, because MockedProvider requires an
      // instance for each query made and the original query is refetched after updates.
      <MockedProvider mocks={[queryMock, queryMock, updateMock]}>
        <BerthEditForm berthId="a" onSubmit={onUpdateMock} pierOptions={pierOptions} />
      </MockedProvider>
    );
    await waitForContent(wrapper);
    await act(async () => {
      wrapper.find('form').simulate('submit');
      await waitForExpect(() => {
        wrapper.update();
        expect(onUpdateMock).toBeCalledTimes(1);
        expect(updateMockCalled).toBe(true);
      });
    });
  });
});
