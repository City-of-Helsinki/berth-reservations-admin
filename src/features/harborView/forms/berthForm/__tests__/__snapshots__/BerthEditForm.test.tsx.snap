// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/harborView/BerthEditForm initially renders loading spinner 1`] = `
<div
  class="spinnerWrapper"
>
  <div
    class="LoadingSpinner-module_loadingSpinner__2xHT_"
  >
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`features/harborView/BerthEditForm renders content after loading 1`] = `
<form
  class="form"
>
  <div
    class="heading"
  >
    <h4
      class="text brand h4"
    >
      Venepaikka
    </h4>
    <button
      type="button"
    >
      <span
        class="text critical span"
      >
        Poista venepaikka
      </span>
    </button>
  </div>
  <div
    class="grid cols3 grid"
  >
    <div
      class="Select-module_root__Ka5uO Select-module_disabled__3MKDP"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="pierId-toggle-button"
        id="pierId-label"
      >
        Laituri
        <span
          class="RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs"
        >
           *
        </span>
      </label>
      <div
        class="Select-module_wrapper__1WQXs"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="pierId-label pierId-toggle-button"
          aria-owns="pierId-menu"
          class="Select-module_button__1aIsm"
          disabled=""
          id="pierId-toggle-button"
          type="button"
        >
          <span
            class="Select-module_buttonLabel__1fqu5"
          >
            A
          </span>
          <svg
            aria-hidden="true"
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Select-module_angleIcon__2-9AD"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h24v24H0z"
              />
              <path
                d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                fill="currentColor"
              />
            </g>
          </svg>
        </button>
        <ul
          aria-labelledby="pierId-label"
          aria-required="true"
          class="Select-module_menu__1H2aU"
          id="pierId-menu"
          role="listbox"
          style="max-height: 260px;"
          tabindex="-1"
        />
      </div>
    </div>
    <div
      class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="number"
      >
        Paikka
      </label>
      <div
        class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
      >
        <input
          class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
          disabled=""
          id="number"
          type="text"
          value="2"
        />
      </div>
    </div>
    <div />
    <div
      class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="width"
      >
        Leveys (m)
      </label>
      <div
        class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
      >
        <input
          class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
          id="width"
          type="text"
          value="2,25"
        />
      </div>
    </div>
    <div
      class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="length"
      >
        Pituus (m)
      </label>
      <div
        class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
      >
        <input
          class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
          id="length"
          type="text"
          value="5"
        />
      </div>
    </div>
    <div
      class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="depth"
      >
        Syvyys (m)
      </label>
      <div
        class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
      >
        <input
          class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
          id="depth"
          type="text"
          value="1"
        />
      </div>
    </div>
  </div>
  <hr />
  <div
    class="Select-module_root__Ka5uO"
  >
    <label
      class="FieldLabel-module_label__1zrXK "
      for="mooringType-toggle-button"
      id="mooringType-label"
    >
      Kiinnitystapa
      <span
        class="RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs"
      >
         *
      </span>
    </label>
    <div
      class="Select-module_wrapper__1WQXs"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="mooringType-label mooringType-toggle-button"
        aria-owns="mooringType-menu"
        class="Select-module_button__1aIsm"
        id="mooringType-toggle-button"
        type="button"
      >
        <span
          class="Select-module_buttonLabel__1fqu5"
        >
          Aisapaikka
        </span>
        <svg
          aria-hidden="true"
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Select-module_angleIcon__2-9AD"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
      <ul
        aria-labelledby="mooringType-label"
        aria-required="true"
        class="Select-module_menu__1H2aU Select-module_overflow__EZdsW"
        id="mooringType-menu"
        role="listbox"
        style="max-height: 260px;"
        tabindex="-1"
      />
    </div>
  </div>
  <div
    class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq"
  >
    <label
      class="FieldLabel-module_label__1zrXK "
      for="comment"
    >
      Huomiot
    </label>
    <div
      class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
    >
      <input
        class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
        id="comment"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="Checkbox-module_checkbox__3L0GR checkbox_hds-checkbox__9HMCz checkbox"
  >
    <input
      checked=""
      class="Checkbox-module_input__3VZvy checkbox_hds-checkbox__input__1w0pu"
      id="isActive"
      type="checkbox"
      value=""
    />
    <label
      class="Checkbox-module_label__L5AN1 checkbox_hds-checkbox__label__3HoD3"
      for="isActive"
    >
      Paikka on käytössä
    </label>
  </div>
  <div
    class="formActionButtons"
  >
    <button
      class="Button-module_button__1msFE button_hds-button__2A0je Button-module_secondary__1nABp button_hds-button--secondary__1NOWS Button-module_theme-coat__2_3U7 button_hds-button--theme-coat__12cO0"
      color="supplementary"
      type="button"
    >
      <span
        class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
      >
        Peruuta
      </span>
    </button>
    <button
      class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO Button-module_theme-coat__2_3U7 button_hds-button--theme-coat__12cO0"
      type="submit"
    >
      <span
        class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
      >
        Tallenna
      </span>
    </button>
  </div>
</form>
`;
